### ISSUE TITLE
(optionally) tear down resources on Tilt exit

### ISSUE THREAD

**Issue by maiamcc** [windmilleng/tilt#2427]
Many users like the fact that exiting Tilt leaves all your old resources up--your next Tilt run can hook into pods from last time (if nothing has changed) and speed up your startup time.

However, for users who want Tilt to tear down resources on exit (as e.g. Docker Compose does), this should be an option. (`tilt up --teardown=true`?)

**Comment by kvokka** [comment]
I like the idea, but maybe we can adopt Docker flag `--rm` here?

in Tilefile it would be reasonable to introduce some `settings` or `defaults` command, where user can overlap it with `rm=true`. This should be an answer for further Tilt configuration switches.

thoughts?

**Comment by maiamcc** [comment]
Yeah, we're already thinking of ways for long-lived configurations to interact with command line flags (so, e.g., you could EITHER set `--rm` as a default somewhere in a config file, OR if you only use it infrequently, pass it as an arg to `tilt up` whenever you need it).

Cool, thanks for all your input! This feature isn't going to be implemented in the immediate future, but a few people have mentioned it already, so it's good to have a central ticket for it. Watch this space for any news, and let us know if your use case changes/if this feature gets any more significant for you.

**Comment by stecky** [comment]
+1 for this enhancement. I do local development on my laptop, but it's nice to have an easy way to shut everything down when I'm not using it. Especially if I am running on battery.

**Comment by nickvollmar** [comment]
Nowadays I run Tilt as `tilt up && tilt down`, which achieves automated teardown behavior (at the cost of reading the Tiltfile twice).

**Comment by tmc** [comment]
FWIW I like the in-tiltfile option here as i think it's the most flexible

**Comment by kevin-lindsay-1** [comment]
IMO it would be nice to have an in-tiltfile approach for this, but in general I think that the CLI invocation itself should be respected above all.

You could say `cleanup/remove/teardown: true` in `tilt_config.json`, create a globally scoped function like `teardown(true)`, but the manual process of providing a flag on the CLI would IMO be the determining factor, because I actually wrote it. That way, I can say "teardown by default, sure, but in this case don't".

---

To piggyback off the previous paragraph, it brings up a question that I think may continually be asked, that being "what are the options for config (outside of custom stuff), and what is the order in which they override?".

In my mind, the order of override would be as follows:
- Declarative config (a la `tilt_config.json`) (least respected)
- CLI arguments
- Hard-coded Tiltfile (most respected)

The issue of best practices comes into play when having an option such as `teardown` in all 3 contexts, leading to the ability to erroneously hard-code `teardown(true)` in `Tiltfile`, ignoring both config and CLI arguments.

**Comment by daniel-kun** [comment]
I second @kevin-lindsay-1 's approach of having the three-tier config options (which must be very properly documented!). I know it from other tools and have had a good experience with it so far.

Personally I would think that when you like the tear-down-behaviour, it must not be the case that all your teammates like it, so a user-config would be a better approach to have this be the default behaviour.

However, when you want to apply it as some kind of "policy" for your whole team, the `Tiltfile` would be a better approach.

Lastly, there surely are unregular cases where you want a different behaviour than that what is configured in your personal config or the Tiltfile.

That's why I would re-order the priority to 
- User config (least respected)
- Tiltfile (overwrites user config)
- CLI arguments (overwrites Tiltfile and user config)

Greetings!
Daniel

**Comment by tclasen** [comment]
+1 from me. I have microservice developers who don't want to know that they are using k8s under the hood, and because of that don't directly use k8s commands through kubectl or even tilt. For these developers I'd like for them to be able to run a single long-running command that stands up their stack and do dev work, and when they close that long-running command it will clean itself up.

Currently I do this through a shell script that hooks the PID of tilt, and when it is done running it issues a `kubectl delete` command on the namespace.

**Comment by leefernandes** [comment]
+1 Using tilt for a development environment a common pain point is explaining that Ctrl+C does not tear down the cluster.  It would be preferable for this to be configurable in the Tiltfile itself.
